src/Starcounter.Database.Extensions/PostCommitTransactor.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Starcounter.Database.ChangeTracking;
+
+namespace Starcounter.Database.Extensions
+{
+    public class PostCommitTransactor : TransactorBase
+    {
+        readonly AsyncLocal<List<KeyValuePair<Type, Change>>> _lastChanges = new AsyncLocal<List<KeyValuePair<Type, Change>>>();
+        readonly PostCommitOptions _hookOptions;
+
+        public PostCommitTransactor(ITransactor transactor, IOptions<PostCommitOptions> postCommitHookOptions)
+            : base(transactor)
+            => _hookOptions = postCommitHookOptions.Value;
+
+        protected override void LeaveContext(IDatabaseContext db, bool exceptionThrown)
+        {
+            List<KeyValuePair<Type, Change>> changes = new List<KeyValuePair<Type, Change>>();
+
+            foreach (var change in db.ChangeTracker.Changes.Where(c => c.Type != ChangeType.Delete))
+            {
+                var proxy = db.Get<object>(change.Oid);
+                var realType = proxy.GetType().BaseType;
+
+                changes.Add(new KeyValuePair<Type, Change>(realType, change));
+            }
+
+            if (changes.Any())
+            {
+                _lastChanges.Value = changes;
+            }
+        }
+
+        protected override void LeftContext()
+        {
+            var changes = _lastChanges.Value;
+
+            if (changes == null)
+            {
+                return;
+            }
+
+            _lastChanges.Value = null;
+
+            RunTask(() =>
```

Comment:
Why async?

Reply:
To leave from the `transactor.Transact` call as soon as possible.

Reply:
I also had a concern about that, if we really need that. The extra context switch take resources too, and the iteration of delegates shouldn't be very heavy I think.

Reply:
Why is it desired? AFAICT, by the time of this call we've left a real transaction and released a database context.

Reply:
My idea was to treat post commit hooks are background tasks, which shall not slow down all application's `ITransactor.Transact` calls.

It's, of course, possible to manually execute heavy operations triggered by a post commit hook in async way, but I thought to make it automatically async in the decorator itself.

Reply:
Such implementation makes certain things impossible to achieve - like ensuring that notifications from serialized transactions are also serialized. I think we don't have to be so opinionated about user requirements.

WTBS, running entire `ExecutePostCommitHooks` asynchronously may have better performance than offloading notifications to a thread pool one-by-one. It it ever becomes an issue we can add a configuration setting in the transactor (RunNorificationsAsynchronously).

Reply:
Okay, made it be fully synchronous, if an async background task is required, it can be scheduled by the app itself.

Reply:
> Okay, made it be fully synchronous

I noticed. Was it decided by the two of you somewhere? 

> if an async background task is required, it can be scheduled by the app itself.

But with more cermony, and to the price that all transaction scopes will include the time it takes to execute post-commit hooks.

I'll let you decide, but I think that the default should be to offload once, the entire scope of invoking callbacks, and allow as a customization to tweak this behavior, with a custom scheduler likely.

Reply:
>... I think that the default should be to offload once ...

It sound reasonable to me, but it complicates the code and testing, also I don't have as much expertise as @bigwad when it comes to asynchronicity and performance, thus I tend to agree with him.